### PR TITLE
gh-105237: Allow calling `issubclass(X, typing.Protocol)` again

### PR DESCRIPTION
**Pull request by AlexWaygood**
* Issue: gh-105237

### CODE REVIEW

File: Lib/test/test_typing.py [hunk 1]
```diff
@@ -2758,6 +2758,16 @@ def x(self): ...
...
 
+        for proto in P, PG, BadP, BadPG:
+            with self.subTest(proto=proto.__name__):
+                self.assertIsSubclass(proto, Protocol)
+
+        # TODO: the behaviour of some of these is questionable!
+        # For now, just check that these don't raise any Exceptions
+        issubclass(object, Protocol)
+        issubclass(str, Protocol)
+        issubclass(C, Protocol)
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]: Some of these appear to be evaluating to `True`, which doesn't seem correct. I didn't investigate this too much, however, as this appears to be longstanding behaviour that I can reproduce on Python 3.8-3.11. So it's not relevant to the regression we're fixing here.

Comment (AlexWaygood) [reply Lib/test/test_typing.py]: Although, perhaps there's not much point in fixing this regression if the results of `issubclass(X, Protocol)` can't be depended upon.

Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: Which "some of these" is this referring to?

```
Python 3.11.0 (v3.11.0:deaf509e8f, Oct 24 2022, 14:43:23) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from typing import Protocol
>>> issubclass(object, Protocol)
False
>>> issubclass(str, Protocol)
False
>>> class C: pass
... 
>>> issubclass(C, Protocol)
False
```

Comment (AlexWaygood) [reply Lib/test/test_typing.py]: Try running this test file:

<details>

```py
import unittest

from typing import TypeVar, runtime_checkable, Protocol

class ProtocolTest(unittest.TestCase):
    def assertIsSubclass(self, cls, class_or_tuple, msg=None):
        if not issubclass(cls, class_or_tuple):
            message = '%r is not a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)

    def assertNotIsSubclass(self, cls, class_or_tuple, msg=None):
        if issubclass(cls, class_or_tuple):
            message = '%r is a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)
    
    def test_protocols_issubclass(self):
        T = TypeVar('T')

        @runtime_checkable
        class P(Protocol):
            def x(self): ...

        @runtime_checkable
        class PG(Protocol[T]):
            def x(self): ...

        class BadP(Protocol):
            def x(self): ...

        class BadPG(Protocol[T]):
            def x(self): ...

        class C:
            def x(self): ...

        self.assertIsSubclass(C, P)
        self.assertIsSubclass(C, PG)
        self.assertIsSubclass(BadP, PG)

        no_subscripted_generics = (
            "Subscripted generics cannot be used with class and instance checks"
        )

        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(C, PG[T])
        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(C, PG[C])

        only_runtime_checkable_protocols = (
            "Instance and class checks can only be used with "
            "@runtime_checkable protocols"
        )

        with self.assertRaisesRegex(TypeError, only_runtime_checkable_protocols):
            issubclass(C, BadP)
        with self.assertRaisesRegex(TypeError, only_runtime_checkable_protocols):
            issubclass(C, BadPG)

        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(P, PG[T])
        with self.assertRaisesRegex(TypeError, no_subscripted_generics):
            issubclass(PG, PG[int])

        only_classes_allowed = r"issubclass\(\) arg 1 must be a class"

        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, P)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, PG)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, BadP)
        with self.assertRaisesRegex(TypeError, only_classes_allowed):
            issubclass(1, BadPG)

        self.assertNotIsSubclass(object, Protocol)
        self.assertNotIsSubclass(str, Protocol)
        self.assertNotIsSubclass(C, Protocol)
        for proto in P, PG, BadP, BadPG:
            with self.subTest(proto=proto.__name__):
                self.assertIsSubclass(proto, Protocol)


if __name__ == "__main__":
    unittest.main()
```

</details>

I also can't reproduce it in the REPL, which makes me think it's probably something cursed to do with a cache somewhere again.

File: Lib/typing.py [hunk 2]
```diff
@@ -1862,8 +1862,6 @@ def meth(self) -> T:
                 ...
     """
     __slots__ = ()
-    _is_protocol = True
```
Comment (JelleZijlstra) [inline Lib/typing.py]: Why are these removed? I bet they're pretty commonly used for introspection.

Comment (AlexWaygood) [reply Lib/typing.py]: Because we didn't need them, and it seemed cleaner without them. But I agree that we shouldn't unnecessarily break users. Added them back in 36b2ac9.

File: Lib/test/test_typing.py (hunk 1 above)
Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: Indeed. I simplified it to this:

<details>

```python
import unittest

from typing import runtime_checkable, Protocol

class ProtocolTest(unittest.TestCase):
    def assertNotIsSubclass(self, cls, class_or_tuple, msg=None):
        if issubclass(cls, class_or_tuple):
            message = '%r is a subclass of %r' % (cls, class_or_tuple)
            if msg is not None:
                message += ' : %s' % msg
            raise self.failureException(message)
    
    def test_protocols_issubclass(self):
        @runtime_checkable
        class P(Protocol):
            def x(self): ...

        class C:
            def x(self): ...
        self.assertNotIsSubclass(C, Protocol)


if __name__ == "__main__":
    unittest.main()
```

</details>

On 3.11 this fails, but if you remove the unused Protocol it passes.

Comment (AlexWaygood) [reply Lib/test/test_typing.py]: Oh, I get it. It's because of the abc module iterating through the subclasses again, right?

It sees that `C` is a subclass of `P`, and since `P` is a subclass of `Protocol`, it determines that `C` is therefore a subclass of `Protocol`. Delightful.

Comment (AlexWaygood) [reply Lib/test/test_typing.py]: Even smaller repro:

```pycon
>>> import typing
>>> @typing.runtime_checkable
... class Foo(typing.Protocol): pass  # empty protocol, so `object` is considered a subclass
...
>>> issubclass(object, typing.Protocol)
True
```

Comment (AlexWaygood) [reply Lib/test/test_typing.py]: (I'm working on a fix)

File: Lib/typing.py
Comment (AlexWaygood) [inline Lib/typing.py:1]: An alternative diff that has the same effect is:

```diff
diff --git a/Lib/typing.py b/Lib/typing.py
index f589be7295..c831ba847f 100644
--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -1782,6 +1782,8 @@ def __init__(cls, *args, **kwargs):
             )

     def __subclasscheck__(cls, other):
+        if cls is Protocol:
+            return type.__subclasscheck__(cls, other)
         if not isinstance(other, type):
             # Same error message as for issubclass(1, int).
             raise TypeError('issubclass() arg 1 must be a class')
@@ -1803,6 +1805,8 @@ def __subclasscheck__(cls, other):
     def __instancecheck__(cls, instance):
         # We need this method for situations where attributes are
         # assigned in __init__.
+        if cls is Protocol:
+            return type.__instancecheck__(cls, instance)
         if not getattr(cls, "_is_protocol", False):
             # i.e., it's a concrete subclass of a protocol
             return super().__instancecheck__(instance)
```

**Comment by AlexWaygood** [comment]
I've run the third-party tests we have at `typing_extensions` on a version of this PR. In combination with #105281, this causes no regressions.

**Review by JelleZijlstra — APPROVED** [review]
